Hazel/src/Platform/Windows/WindowsWindow.cpp

```diff
@@ -47,15 +47,29 @@ namespace Hazel {
 			glfwSetErrorCallback(GLFWErrorCallback);
 			s_GLFWInitialized = true;
 		}
-
+		
+		// Create window in windowed mode
+		m_Data.Mode = WindowMode::Windowed;
+		glfwWindowHint(GLFW_DECORATED, GLFW_TRUE);
+		glfwWindowHint(GLFW_RESIZABLE, GLFW_TRUE);
+		glfwWindowHint(GLFW_AUTO_ICONIFY, GLFW_FALSE);
 		m_Window = glfwCreateWindow((int)props.Width, (int)props.Height, m_Data.Title.c_str(), nullptr, nullptr);
 
 		m_Context = new OpenGLContext(m_Window);
 		m_Context->Init();
 
 		glfwSetWindowUserPointer(m_Window, &m_Data);
-		SetVSync(true);
 
+		int Xpos, Ypos;
+		glfwGetWindowPos(m_Window, &Xpos, &Ypos);
+		m_Data.WindowedPos = { Xpos, Ypos };
+
+		SetVSync(props.VSync);
+
+		// Change mode if needed
+		if (m_Data.Mode != props.Mode)
+			SetWindowMode(props.Mode);
```

Comment:
This line could result in the `WindowMode` (and window dimensions) changing before any callbacks are registered and also before the `EventCallbackFn` is set. This means that this WindowResizeEvent is lost. Because the EventCallbackFn is only set after construction there is actually no way for us to emit any events during this code block (unless the EventCallbackFn becomes a parameter instead of being set afterwards).

This also results in `m_Data.Width` and `m_Data.Height` not being initialized and the OpenGL Viewport not being resized (in my implementation). We could manually call the Viewport method after this, but that still leaves the uninitialized variables.

It should also be noted that if the `WindowMode` is not `Windowed`, then the Application has to also get the Window dimensions after creating the Window since it has no immediate way of knowing the Window's new size. This seems like a bit of a clumsy API. The only solution I see is not allowing `WindowMode` as a parameter during Window cosntruction and then having the `SetWindowMode` method return the new window dimensions in a vec2. The application could then just call SetWindowMode after creating the Window.

Reply:
> This line could result in the `WindowMode` (and window dimensions) changing before any callbacks are registered and also before the `EventCallbackFn` is set. This means that this WindowResizeEvent is lost. Because the EventCallbackFn is only set after construction there is actually no way for us to emit any events during this code block (unless the EventCallbackFn becomes a parameter instead of being set afterwards).
>
> This also results in `m_Data.Width` and `m_Data.Height` not being initialized and the OpenGL Viewport not being resized (in my implementation). We could manually call the Viewport method after this, but that still leaves the uninitialized variables.

It is ment to not fire any events, as events are ment to fire after creating the context only. I can see how this is a bit flawed atm. One solution would be to create the context right before the events are registered, after the m_Window is fully created/altered.

I see how `m_Data.Width` and `m_Data.Height` are not set when you stay in Windowed mode, but in borderless/fullscreen this is taken care of in the `SetWindowMode` call.

> It should also be noted that if the `WindowMode` is not `Windowed`, then the Application has to also get the Window dimensions after creating the Window since it has no immediate way of knowing the Window's new size. This seems like a bit of a clumsy API. The only solution I see is not allowing `WindowMode` as a parameter during Window cosntruction and then having the `SetWindowMode` method return the new window dimensions in a vec2. The application could then just call SetWindowMode after creating the Window.

Not realy true, after the Window is constructed (before the body of the application constructor is called), application can call `m_Window.GetWidth()` and `.GetHeight()` to know it's dimensions

My solution would be:
```cpp
		... // same code as we have now
		m_Window = glfwCreateWindow((int)props.Width, (int)props.Height, m_Data.Title.c_str(), nullptr, nullptr);

		int Xpos, Ypos; // Extract the position in windowed mode
		glfwGetWindowPos(m_Window, &Xpos, &Ypos);
		m_Data.WindowedPos = { Xpos, Ypos };

		if (m_Data.Mode == props.Mode) {
			m_Data.Width = props.Width
			m_Data.Height = props.Height
		} else {
			SetWindowMode(props.Mode);
		}

		// Set additional window parameters
		SetVSync(props.VSync);

		// Now the window is constructed, we can create a context for it
		m_Context = new OpenGLContext(m_Window);
		m_Context->Init();

		// Set GLFW callbacks
		glfwSetWindowUserPointer(m_Window, &m_Data);

		... // same code as we have now
```

What do you think about this @Puddlestomper ?

Reply:
Resolved in https://github.com/TheCherno/Hazel/pull/108/commits/13b0aca5316d229b6b04faf820ec658b05892392